ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -289,85 +291,104 @@ private class ContractsFetch(
+                )(lc)
+                (stepsAndOffset.out0, stepsAndOffset.out1)
+
+              case (AbsoluteBookmark(_), _) | (LedgerBegin, true) =>
+                val stepsAndOffset = builder add transactionsFollowingBoundary(txnK)
+                stepsAndOffset.in <~ Source.single(startOffset)
+                (
+                  (stepsAndOffset: FanOutShape2[_, ContractStreamStep.LAV1, _]).out0,
+                  stepsAndOffset.out1,
+                )
+            }
 
-          val transactInsertsDeletes = Flow
-            .fromFunction(
-              jsonifyInsertDeleteStep(
-                (_: InsertDeleteStep[Any, lav1.event.CreatedEvent]),
-                templateId,
+            val transactInsertsDeletes = Flow
+              .fromFunction(
+                jsonifyInsertDeleteStep(
+                  (_: InsertDeleteStep[Any, lav1.event.CreatedEvent]),
+                  templateId,
+                )
               )
-            )
-            .via(if (sjd.q.queries.allowDamlTransactionBatching) conflation else Flow.apply)
-            .map(insertAndDelete)
+              .via(if (sjd.q.queries.allowDamlTransactionBatching) conflation else Flow.apply)
+              .map(insertAndDelete)
 
-          idses.map(_.toInsertDelete) ~> transactInsertsDeletes ~> acsSink
-          lastOff ~> offsetSink
+            idses.map(_.toInsertDelete) ~> transactInsertsDeletes ~> acsSink
+            lastOff ~> offsetSink
 
-          ClosedShape
-        }
-      )
+            ClosedShape
+          }
+        )
 
-      val (acsQueue, lastOffsetFuture) = graph.run()
+        val (acsQueue, lastOffsetFuture) = graph.run()
+
+        for {
+          _ <- sinkCioSequence_(acsQueue)
+          offset0 <- connectionIOFuture(lastOffsetFuture)
+          offsetOrError <- offset0 max AbsoluteBookmark(absEnd.toDomain) match {
+            case ab @ AbsoluteBookmark(newOffset) =>
+              ContractDao
+                .updateOffset(parties, templateId, newOffset, offsets)
+                .map(_ => ab)
+            case LedgerBegin =>
+              fconn.pure(LedgerBegin)
+          }
+        } yield offsetOrError
+      }
+  }
 
-      for {
-        _ <- sinkCioSequence_(acsQueue)
-        offset0 <- connectionIOFuture(lastOffsetFuture)
-        offsetOrError <- offset0 max AbsoluteBookmark(absEnd.toDomain) match {
-          case ab @ AbsoluteBookmark(newOffset) =>
-            ContractDao
-              .updateOffset(parties, templateId, newOffset, offsets)
-              .map(_ => ab)
-          case LedgerBegin =>
-            fconn.pure(LedgerBegin)
-        }
-      } yield offsetOrError
-    }
+  private def logInteractionLedger[T, C](
+      templateId: domain.ContractTypeId.Resolved
+  )(block: => T)(implicit lc: LoggingContextOf[C]): T = {
+    if (logger.debug.isEnabled) {
+      val startTime = System.nanoTime()
+      logger.debug(s"Starting cache refresh for TemplateId $templateId")
+      val result = block
```

Comment:
We should ensure exception's are reported too, e.g.
```suggestion
      val result = 
        try { 
          block
        } catch {
          e: Exception => {
            val endTime = System.nanoTime()
            logger.debug("Failed $action after ${(endTime - startTime) / 1000000L} ms because $e")
            throw e
          }
        }
```